src/hammer-vlsi/hammer_vlsi/constraints.py

```diff
@@ -242,42 +244,138 @@ class PlacementConstraint(NamedTuple('PlacementConstraint', [
     __slots__ = ()
 
     @staticmethod
-    def from_dict(constraint: dict) -> "PlacementConstraint":
+    def from_dict(masters: List[MacroSize], constraint: dict) -> "PlacementConstraint":
         constraint_type = PlacementConstraintType.from_str(
             str(constraint["type"]))
+
+        ### Margins ###
+        # This field is mandatory in TopLevel constraints
+        # This field is disallowed otherwise
         margins = None  # type: Optional[Margins]
-        orientation = None  # type: Optional[str]
-        top_layer = None  # type: Optional[str]
-        layers = None  # type: Optional[List[str]]
-        obs_types = None  # type: Optional[List[ObstructionType]]
-        if constraint_type == PlacementConstraintType.TopLevel:
+        if "margins" in constraint:
+            assert constraint_type == PlacementConstraintType.TopLevel, "Non-TopLevel constraint must not contain margins: {}".format(constraint)
```

Comment:
Nitpick-ish: could we have these be raised as `ValueError`s? (since we have more precise knowledge than just some random assertion failed)

Reply:
Sure, that's fine. I think at some point we want a more elegant way to handle these than either assertions or exceptions (e.g. you might want to let the tool continue a little bit longer to give you multiple errors before exiting)